### PR TITLE
Add temperature controller sample model

### PR DESCRIPTION
**Pull request by rido-min**
This PR replaces #61 

The interface is composed with components and root elements. The command uses a complex object response.
DeviceInfo is included to show how to reuse interfaces from the public repo.

**TemperatureController**
- Telelemetry Root: WorkingSet
- Command Root: Reboot
- Read Prop Root. Serial Number
- Component Thermostat: thermostat1
  - Telemetry. temperature
  - Writable Prop. targetTemperature
  - Read Prop. maxTempSinceLastReboot
  - Command. GetMinMaxReport
- Component Thermostat: thermostat2
  - Telemetry. temperature
  - Writable Prop. targetTemperature
  - Read Prop. maxTempSinceLastReboot
  - Command. GetMinMaxReport
- Component. [DeviceInfo](https://repo.azureiotrepository-test.com/Models/dtmi:azure:DeviceManagement:DeviceInformation;1?api-version=2020-05-01-preview)

### CODE REVIEW

File: DTDL/v2/samples/TemperatureController.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  "contents": [    {
+      "@type": "Telemetry",
+      "name": "workingSet",
+      "displayName": "Working Set",
+      "description": "Current working set of the device memory.",
+      "schema": "double"
+    },
+    {
+      "@type": "Property",
+      "name": "serialNumber",
```
Comment (jspaith) [inline DTDL/v2/samples/TemperatureController.json:16]: Nit: Should this have a displayName and description for consistency with other fields?

File: DTDL/v2/samples/TemperatureController.json [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  "@context": "dtmi:dtdl:context;2",
+  "@id": "dtmi:com:example:TemperatureController;1",
+  "@type": "Interface",
+  "displayName": "Temperature Controller",
+  "description": "Device with two thermostats and remote reboot.",
+  "contents": [    {
+      "@type": "Telemetry",
+      "name": "workingSet",
+      "displayName": "Working Set",
+      "description": "Current working set of the device memory.",
```
Comment (jspaith) [inline DTDL/v2/samples/TemperatureController.json:11]: Please specify units, e.g. KB.

File: DTDL/v2/samples/TemperatureController.json (hunk 1 above)
Comment (jspaith) [reply DTDL/v2/samples/TemperatureController.json:16]: Similiar feedback about description for themostat1/2 fields below.  I don't really care but want to make sure it's conscious decision either way.

File: DTDL/v2/samples/TemperatureController.json [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      "displayName": "Thermostat One"
+    },
+    {
+      "@type" : "Component",
+      "schema": "dtmi:com:example:Thermostat;1",
+      "name": "thermostat2",
+      "displayName": "Thermostat Two"
+    },
+    {
+      "@type": "Component",
```
Comment (jspaith) [inline DTDL/v2/samples/TemperatureController.json:44]: Please remove this (for now at least) along with the DTDL of this file itself until some other areas get agreement.

File: DTDL/v2/samples/TemperatureController.json (hunk 2 above)
Comment (rido-min) [reply DTDL/v2/samples/TemperatureController.json]: done, also using `DataSize` semantic type

File: DTDL/v2/samples/TemperatureController.json [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  "description": "Device with two thermostats and remote reboot.",
+  "contents": [    {
+      "@type": [
+        "Telemetry", "DataSize"
+      ],
+      "name": "workingSet",
+      "displayName": "Working Set",
+      "description": "Current working set of the device memory in KB.",
+      "schema": "double",
+      "unit" : "kibibyte"
```
Comment (ericmitt) [inline DTDL/v2/samples/TemperatureController.json:15]: Kilobyte, typo

Comment (rido-min) [reply DTDL/v2/samples/TemperatureController.json:15]: it's not a typo, using values from https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/dtdlv2.md#semantic-types

File: DTDL/v2/samples/DeviceInfo.json [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    "@id": "dtmi:azure:DeviceManagement:DeviceInformation;1",
+    "@type": "Interface",
+    "displayName": "Device Information",
+    "contents": [
+        {
+            "@type": "Property",
+            "name": "manufacturer",
+            "displayName": "Manufacturer",
+            "schema": "string",
+            "description": "Company name of the device manufacturer. This could be the same as the name of the original equipment manufacturer (OEM). Ex. Contoso."
```
Comment (briancr-ms) [inline DTDL/v2/samples/DeviceInfo.json:11]: >Ex. [](start = 150, length = 3)

I wonder if we should spell out the word "Example" in all of these descriptions? It's not clear to me that "Ex." is a known abbreviation for "example".

File: DTDL/v2/samples/DeviceInfo.json [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        },
+        {
+            "@type": "Property",
+            "name": "processorManufacturer",
+            "displayName": "Processor manufacturer",
+            "schema": "string",
+            "description": "Name of the manufacturer of the processor on the device. Ex. Intel."
+        },
+        {
+            "@type": "Property",
```
Comment (briancr-ms) [inline DTDL/v2/samples/DeviceInfo.json:49]: >"@type": "Property" [](start = 12, length = 19)

Should we consider adding the DataSize semantic type here and the appropriate unit (probably kibibyte)?

File: DTDL/v2/samples/DeviceInfo.json [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+            "name": "totalStorage",
+            "displayName": "Total storage",
+            "schema": "double",
+            "description": "Total available storage on the device in kilobytes. Ex. 2048000 kilobytes."
+        },
+        {
+            "@type": "Property",
```
Comment (briancr-ms) [inline DTDL/v2/samples/DeviceInfo.json:56]: >"@type": "Property" [](start = 12, length = 19)

Should we consider adding the DataSize semantic type here and the appropriate unit (probably kibibyte)?

File: DTDL/v2/samples/TemperatureController.json (hunk 4 above)
Comment (briancr-ms) [reply DTDL/v2/samples/TemperatureController.json:15]: "kibibyte" is the correct unit name for 1,024 bytes.

---
In reply to: [445771042](https://github.com/Azure/opendigitaltwins-dtdl/pull/62#discussion_r445771042) [](ancestors = 445771042)

File: DTDL/v2/samples/TemperatureController.json [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+  "@type": "Interface",
+  "displayName": "Temperature Controller",
+  "description": "Device with two thermostats and remote reboot.",
+  "contents": [    {
+      "@type": [
+        "Telemetry", "DataSize"
+      ],
+      "name": "workingSet",
+      "displayName": "Working Set",
+      "description": "Current working set of the device memory in KB.",
```
Comment (briancr-ms) [inline DTDL/v2/samples/TemperatureController.json:13]: >KB [](start = 66, length = 2)

Since the unit is kibibyte, the unit symbol should be KiB (not KB).

File: DTDL/v2/samples/Thermostat.json [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  "description": "Reports current temperature and provides desired temperature control.",
+  "contents": [
+    {
+      "@type": [
+        "Telemetry",
+        "Temperature"
+      ],
+      "name": "temperature",
+      "displayName" : "Temperature",
+      "description" : "Temperature in celsius.",
```
Comment (briancr-ms) [inline DTDL/v2/samples/Thermostat.json:15]: >celsius [](start = 38, length = 7)

Celsius should be capitalized (it's a name). It's usually written "degrees Celsius", so this should be "Temperature in degrees Celsius."

File: DTDL/v2/samples/Thermostat.json [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+      "schema": "double",
+      "unit" : "degreeCelsius",
+      "displayName": "Max temperature since last reboot.",
+      "description": "Returns the max temperature since last device reboot."
+    },
+    {
+      "@type": "Command",
+      "name": "getMaxMinReport",
+      "displayName": "Get Max-Min report.",
+      "description": "This command returns the max, min and average temperature in the specified period.",
```
Comment (briancr-ms) [inline DTDL/v2/samples/Thermostat.json:46]: >in the specified period. [](start = 80, length = 24)

The description says that the command returns the data in the specified period, but, from the command request, it looks like the command returns the data from the time specified to the current time. Perhaps reword this as "...from the specified time to the current time." (Or, if the intent is to support a time span, the request should support a start and end time.)

File: DTDL/v2/samples/Thermostat.json [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        "displayName": "Since",
+        "description": "Period to return the max-min report.",
+        "schema": "dateTime"
+      },
+      "response": {
+        "name" : "tempReport",
+        "displayName": "Temperature Report",
+        "schema": {
+          "@type": "Object",
+          "fields": [
```
Comment (briancr-ms) [inline DTDL/v2/samples/Thermostat.json:58]: Does it make sense to include the end time in the report, since the response may take some time to get back to the caller or it may be stored?

File: DTDL/v2/samples/TemperatureController.json (hunk 3 above)
Comment (briancr-ms) [reply DTDL/v2/samples/TemperatureController.json:51]: It looks like this comment is from iteration 2, but DeviceInfo is still in this PR. Is it being removed?

---
In reply to: [445206712](https://github.com/Azure/opendigitaltwins-dtdl/pull/62#discussion_r445206712) [](ancestors = 445206712)

Comment (rido-min) [reply DTDL/v2/samples/TemperatureController.json:51]: I removed, and added it back again. The reason is to use an interface from the global repo, that will help to validate resolution scenarios

File: DTDL/v2/samples/DeviceInfo.json (hunk 5 above)
Comment (rido-min) [reply DTDL/v2/samples/DeviceInfo.json:11]: This PR is not about deviceInfo, I just downloaded the one from the global repo

File: DTDL/v2/samples/DeviceInfo.json (hunk 6 above)
Comment (rido-min) [reply DTDL/v2/samples/DeviceInfo.json:49]: yes, but not in this PR

File: DTDL/v2/samples/TemperatureController.json (hunk 8 above)
Comment (rido-min) [reply DTDL/v2/samples/TemperatureController.json:13]: from https://en.wikipedia.org/wiki/Kibibyte seems like the units are equivalent between IEC and JEDEC
image

File: DTDL/v2/samples/Thermostat.json (hunk 10 above)
Comment (rido-min) [reply DTDL/v2/samples/Thermostat.json:46]: fixed

File: DTDL/v2/samples/Thermostat.json (hunk 11 above)
Comment (rido-min) [reply DTDL/v2/samples/Thermostat.json:58]: yep, added.

File: DTDL/v2/samples/Thermostat.json (hunk 9 above)
Comment (rido-min) [reply DTDL/v2/samples/Thermostat.json:15]: fixed

File: DTDL/v2/samples/DeviceInfo.json (hunk 7 above)
Comment (rido-min) [reply DTDL/v2/samples/DeviceInfo.json:56]: yes but not in this PR

**Comment by rido-min** [comment]
@briancr-ms I'm adding the DeviceInfo interface -copied from the public repo- in this PR to enable DTDL parser validation.

Do we want to make this repo the home for common interfaces?

File: DTDL/v2/samples/TemperatureController.json (hunk 8 above)
Comment (briancr-ms) [reply DTDL/v2/samples/TemperatureController.json:13]: I read this table as saying that the unit symbol for kibibyte is KiB and the symbol for kilobyte is KB. Since we are using the standard unit kibibyte (not kilobyte), I believe we should use the symbol KiB.

---
In reply to: [445814446](https://github.com/Azure/opendigitaltwins-dtdl/pull/62#discussion_r445814446) [](ancestors = 445814446)

File: DTDL/v2/samples/DeviceInfo.json (hunk 5 above)
Comment (briancr-ms) [reply DTDL/v2/samples/DeviceInfo.json:11]: Why are we adding DeviceInfo in the samples folder then? We only want one copy of DeviceInfo. If it's not a sample, but a real interface, then let's put it somewhere appropriate.

---
In reply to: [445812764](https://github.com/Azure/opendigitaltwins-dtdl/pull/62#discussion_r445812764) [](ancestors = 445812764)

File: DTDL/v2/samples/TemperatureController.json (hunk 8 above)
Comment (rido-min) [reply DTDL/v2/samples/TemperatureController.json:13]: I'm worried of adding a new term a few people might know. This is the first time me or @ericmitt heard about it.

File: DTDL/v2/samples/DeviceInfo.json (hunk 5 above)
Comment (rido-min) [reply DTDL/v2/samples/DeviceInfo.json:11]: .. to be able to validate the DTDL with the current parser.

Comment (briancr-ms) [reply DTDL/v2/samples/DeviceInfo.json]: But all the definitions don't need to be in the same folder to be validated with the DTDL v2 parser. They can come from anywhere. The parser will call your client back when it comes across a model reference that it doesn't have. Then you can provide it (getting it from wherever you want).

---
In reply to: [445828794](https://github.com/Azure/opendigitaltwins-dtdl/pull/62#discussion_r445828794) [](ancestors = 445828794)

File: DTDL/v2/samples/Thermostat.json [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+              "schema": "double"
+            },
+            {
+              "name" : "avgTemp",
+              "displayName": "Average Temperature",
+              "schema": "double"
+            },
+            {
+              "name" : "startTime",
+              "displayName": "Since",
```
Comment (briancr-ms) [inline DTDL/v2/samples/Thermostat.json:76]: >Since [](start = 30, length = 5)

Display name should be "Start time".

File: DTDL/v2/samples/Thermostat.json [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+              "schema": "dateTime"
+            },
+            {
+              "name" : "endTime",
+              "displayName": "Since",
```
Comment (briancr-ms) [inline DTDL/v2/samples/Thermostat.json:81]: >Since [](start = 30, length = 5)

Display name should be "End time".

File: DTDL/v2/samples/Thermostat.json (hunk 12 above)
Comment (rido-min) [reply DTDL/v2/samples/Thermostat.json]: fixed

File: DTDL/v2/samples/Thermostat.json (hunk 13 above)
Comment (rido-min) [reply DTDL/v2/samples/Thermostat.json]: fixed

File: DTDL/v2/samples/DeviceInfo.json (hunk 7 above)
Comment (ericmitt) [reply DTDL/v2/samples/DeviceInfo.json]: Agree, not in this PR
